man/man1/logger.1
.\" @(#)logger.1 1.1 92/07/30 SMI; from UCB 4.3
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.TH LOGGER 1 "9 September 1987"
.SH NAME
logger \- add entries to the system log
.SH SYNOPSIS
.B logger
[
.B \-t
.I tag
] [
.B \-p
.I priority
] [
.B \-i
] [
.B \-f
.I filename
] [
.I message
] .\|.\|.
.SH DESCRIPTION
.IX  "logger command"  ""  "\fLlogger\fP \(em make system log entry"
.IX  make "system log entry \(em \fLlogger\fP"
.IX  create "system log entry \(em \fLlogger\fP"
.B logger
provides a method for adding one-line entries to the system log file
from the command line.  One or more
.I message
arguments can be given on the command line, in which case each is
logged immediately.  Otherwise, a
.I filename
can be specified, in which case each line in the file
is logged.  If neither is specified,
.B logger
reads and logs messages on a line-by-line basis from the standard input.
.SH OPTIONS
.TP 12n
.BI \-t \ tag
Mark each line added to the log with the specified
.IR tag .
.TP
.BI \-p \ priority
Enter the message with the specified
.IR priority .
The message priority can be specified numerically, or as a
.IB facility . level
pair.  For example,
.RB ` "\-p local3.info" '
assigns the message priority to the
.B info
level in the
.B local3
facility.  The default priority is
.BR user.notice .
.TP
.B \-i
Log the process
.SM ID
of the
.B logger
process with each line.
.TP
.BI \-f \ filename
Use the contents of
.I filename
as the message to log.
.TP
.I message
If this is unspecified, either the file
indicated with
.B \-f
or the standard input is added to the log.
.SH EXAMPLES
.RS
.ft B
logger System rebooted
.ft R
.RE
.LP
will log the message
.RB ` "System rebooted" '
to the facility at priority
.B notice
to be treated by
.B syslogd
as other messages to the facility
.B notice
are.
.IP
.B
logger \-p local0.notice \-t \s-1HOSTIDM\s0 \-f /dev/idmc
.LP
will read from the file
.B /dev/idmc
and will log each line in that file as a message with the tag
.RB ` \s-1HOSTIDM\s0 '
at priority
.B notice
to be treated by
.B syslogd
as other messages to the facility
.B local0
are.
.SH SEE ALSO
.BR syslog (3),
.BR syslogd (8)


ucb/logger.c
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1983 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

#ifndef lint
static char sccsid[] = "@(#)logger.c 1.1 92/07/30 SMI"; /* from UCB 6.2 9/19/85 */
#endif not lint

#include <stdio.h>
#include <syslog.h>
#include <ctype.h>

/*
**  LOGGER -- read and log utility
**
**	This routine reads from an input and arranges to write the
**	result on the system log, along with a useful tag.
*/

main(argc, argv)
	int argc;
	char **argv;
{
	char buf[200];
	char *tag;
	register char *p;
	int pri = LOG_NOTICE;
	int logflags = 0;
	extern char *getlogin();

	/* initialize */
	tag = getlogin();

	/* crack arguments */
	while (--argc > 0)
	{
		p = *++argv;
		if (*p != '-')
			break;

		switch (*++p)
		{
		  case '\0':		/* dummy */
			/* this can be used to give null parameters */
			break;

		  case 't':		/* tag */
			if (argc > 1 && argv[1][0] != '-')
			{
				argc--;
				tag = *++argv;
			}
			else
				tag = NULL;
			break;

		  case 'p':		/* priority */
			if (argc > 1 && argv[1][0] != '-')
			{
				argc--;
				pri = pencode(*++argv);
			}
			break;

		  case 'i':		/* log process id also */
			logflags |= LOG_PID;
			break;

		  case 'f':		/* file to log */
			if (argc > 1 && argv[1][0] != '-')
			{
				argc--;
				if (freopen(*++argv, "r", stdin) == NULL)
				{
					fprintf("logger: ");
					perror(*argv);
					exit(1);
				}
			}
			break;

		  default:
			fprintf(stderr, "logger: unknown flag -%s\n", p);
			break;
		}
	}

	/* setup for logging */
	openlog(tag, logflags, 0);
	(void) fclose(stdout);

	/* log input line if appropriate */
	if (argc > 0)
	{
		char buf[120];

		buf[0] = '\0';
		while (argc-- > 0)
		{
			strcat(buf, " ");
			strcat(buf, *argv++);
		}
		syslog(pri, "%s", buf + 1);
		exit(0);
	}

	/* main loop */
	while (fgets(buf, sizeof buf, stdin) != NULL)
		syslog(pri, "%s", buf);

	exit(0);
	/* NOTREACHED */
}


struct code {
	char	*c_name;
	int	c_val;
};

struct code	PriNames[] = {
	"panic",	LOG_EMERG,
	"emerg",	LOG_EMERG,
	"alert",	LOG_ALERT,
	"crit",		LOG_CRIT,
	"err",		LOG_ERR,
	"error",	LOG_ERR,
	"warn",		LOG_WARNING,
	"warning",	LOG_WARNING,
	"notice",	LOG_NOTICE,
	"info",		LOG_INFO,
	"debug",	LOG_DEBUG,
	NULL,		-1
};

struct code	FacNames[] = {
	"kern",		LOG_KERN,
	"user",		LOG_USER,
	"mail",		LOG_MAIL,
	"daemon",	LOG_DAEMON,
	"auth",		LOG_AUTH,
	"security",	LOG_AUTH,
	"lpr",		LOG_LPR,
	"news",		LOG_NEWS,
	"uucp",		LOG_UUCP,
	"cron",		LOG_CRON,
	"local0",	LOG_LOCAL0,
	"local1",	LOG_LOCAL1,
	"local2",	LOG_LOCAL2,
	"local3",	LOG_LOCAL3,
	"local4",	LOG_LOCAL4,
	"local5",	LOG_LOCAL5,
	"local6",	LOG_LOCAL6,
	"local7",	LOG_LOCAL7,
	NULL,		-1
};


/*
 *  Decode a symbolic name to a numeric value
 */

pencode(s)
	register char *s;
{
	register char *p;
	int lev;
	int fac;
	char buf[100];

	for (p = buf; *s && *s != '.'; )
		*p++ = *s++;
	*p = '\0';
	if (*s++) {
		fac = decode(buf, FacNames);
		if (fac < 0)
			bailout("unknown facility name: ", buf);
		for (p = buf; *p++ = *s++; )
			continue;
	} else
		fac = 0;
	lev = decode(buf, PriNames);
	if (lev < 0)
		bailout("unknown priority name: ", buf);

	return ((lev & LOG_PRIMASK) | (fac & LOG_FACMASK));
}


decode(name, codetab)
	char *name;
	struct code *codetab;
{
	register struct code *c;
	register char *p;
	char buf[40];

	if (isdigit(*name))
		return (atoi(name));

	(void) strcpy(buf, name);
	for (p = buf; *p; p++)
		if (isupper(*p))
			*p = tolower(*p);
	for (c = codetab; c->c_name; c++)
		if (!strcmp(buf, c->c_name))
			return (c->c_val);

	return (-1);
}

bailout(a, b)
	char *a, *b;
{
	fprintf(stderr, "logger: %s%s\n", a, b);
	exit(1);
}
